solutions/platinum/poi-10-frog.mdx

```diff
@@ -2,96 +2,163 @@
-ll m;
+**Time Complexity:** $\mathcal{O}(N\log M)$
 
-int get(int x, int lef) {
-	F0Rd(i, 20) if (lef & (1 << i)) x = nex[x][i];
-	return x;
-}
-
-void solve(int x) {
-	if (m <= n) cout << get(x, m) << " ";
-	else {
-		x = get(x, n);
-		cout << get(x, (m - n) % sz[x]) << " ";
-	}
-}
+**Space Complexity:** $\mathcal{O}(N)$
 
-ll p[MX];
+## Solution 2
+Observe that a functional graph is constructed through these jumps. After $n$ jumps, any node that
+we start from will end up being inside a cycle. Thus, we can calculate our binary lifts like normal,
+and calculate the number of jumps, modulo the cycle size.
 
-void genNex() {
-	pii cur = {1, k + 1};
-	FOR(i, 1, n + 1) {
-		while (cur.s < n && p[cur.s + 1] - p[i] < p[i] - p[cur.f])
-			cur.f++, cur.s++;
-		if (p[i] - p[cur.f] >= p[cur.s] - p[i]) nex[i][0] = cur.f;
-		else nex[i][0] = cur.s;
-	}
-}
+**Time Complexity:** $\mathcal{O}(N\log N)$
 
-int vis[MX];
+**Space Complexity:** $\mathcal{O}(N\log N)$
 
-void cycSize() {
-	FOR(i, 1, n + 1) {
-		int I = i;
-		for (; vis[I] == 0; I = nex[I][0]) vis[I] = i;
-		if (vis[I] != i) continue;
+## Implementation #1
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
 
-		int cur = 0;
-		for (; vis[I] == i; I = nex[I][0]) {
-			cur++;
-			vis[I] += n;
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	int n, k;
+	ll m;
+	cin >> n >> k >> m;
+	vector<ll> rocks(n);
+	for (ll &i : rocks) { cin >> i; }
+	vector<int> next_stone(n);
+	/*
+	 * Maintain a sliding window of
+	 * size k + 1 to calculate the next stone
+	 * the frog will jump to.
+	 */
+	int l = 0, r = k;
+	for (int i = 0; i < n; i++) {
+		while (r + 1 < n && rocks[r + 1] - rocks[i] < rocks[i] - rocks[l]) {
+			++l, ++r;
 		}
-		for (; vis[I] == i + n; I = nex[I][0]) {
-			sz[I] = cur;
-			vis[I] += n;
+		next_stone[i] = (rocks[i] - rocks[l] >= rocks[r] - rocks[i]) ? l : r;
+	}
+	const int LOG = 64 - __builtin_clzll(m);
+	vector<int> res(n);
+	iota(begin(res), end(res), 0);
+	/*
+	 * Calculate the 2^i-th bit at every step,
+	 * and check if we need to make this "lift".
+	 */
+	for (int i = 0; i < LOG; i++) {
+		if ((m >> i) & 1) {
+			for (int j = 0; j < n; j++) { res[j] = next_stone[res[j]]; }
 		}
+		vector<int> nxt_t(n);
+		for (int j = 0; j < n; j++) { nxt_t[j] = next_stone[next_stone[j]]; }
+		next_stone = nxt_t;
 	}
+	for (int i = 0; i < n; i++) { cout << res[i] + 1 << " \n"[i == n - 1]; }
+	return 0;
 }
+```
+
+## Implementation #2
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
 
 int main() {
-	ios_base::sync_with_stdio(0);
-	cin.tie(0);
+	cin.tie(0)->sync_with_stdio(0);
+	int n, k;
+	ll m;
 	cin >> n >> k >> m;
-	FOR(i, 1, n + 1) cin >> p[i];
-	genNex();
-	FOR(i, 1, 20) FOR(j, 1, n + 1) nex[j][i] = nex[nex[j][i - 1]][i - 1];
-	cycSize();
-	FOR(i, 1, n + 1) solve(i);
+	vector<ll> rocks(n);
+	for (ll &i : rocks) { cin >> i; }
+	const int LOG = 32 - __builtin_clz(n);
+	vector<vector<int>> lift(n, vector<int>(LOG));
```

Comment:
would recommend you just use std log functions: https://en.cppreference.com/w/cpp/numeric/math/log2